File: 299-bulls-and-cows/bulls-and-cows.java
class Solution {
    public String getHint(String secret, String guess) {
        HashMap<Character, Integer> m = new HashMap<>();

        for (char c : secret.toCharArray()) {
            m.put(c, m.getOrDefault(c, 0) + 1);
        }

        int bulls = 0;
        int cows = 0;

        for (int i = 0; i < guess.length(); i++) {
            if (secret.charAt(i) == guess.charAt(i)) {
                bulls += 1;
                char c = secret.charAt(i);
                m.put(c, m.getOrDefault(c, 0) - 1);
            } 
        }
        for (int i = 0; i < guess.length(); i++){
                if ((m.getOrDefault(guess.charAt(i), 0) != 0) && !(secret.charAt(i) == guess.charAt(i))){
                    System.out.println(guess.charAt(i) + " " + m +  " " +m.getOrDefault(guess.charAt(i), 0));
                    cows += 1;
                    char c = guess.charAt(i);
                    m.put(c, m.getOrDefault(c, 0) - 1);
                }
        }
        // System.out.println(bulls + " " + cows);
        return bulls + "A" + cows + "B";
    
    }
}

File: 299-bulls-and-cows/bulls-and-cows.py
class Solution:
    def getHint(self, secret: str, guess: str) -> str:
        bulls = 0
        cows = 0
        secMap = {}
        guMap = {}
        for i in range(len(secret)):
            if secret[i] == guess[i]:
                bulls += 1
            else:
                secMap[secret[i]] = secMap.get(secret[i], 0) + 1
        
        for i in range(len(secret)):
            if secret[i] != guess[i] and secMap.get(guess[i], -1) > 0:
                secMap[guess[i]] -= 1
                cows += 1
        return (f"{bulls}A{cows}B")